### PR TITLE
MPT

### PR DESCRIPTION
**Pull request by shawnxie999**
## High Level Overview of Change
https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-multi-purpose-tokens

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### Did you update HISTORY.md?

- [ ] Yes
- [ ] No, this change does not impact library users

### CODE REVIEW

File: packages/ripple-binary-codec/src/enums/definitions.json
Comment (mvadari) [inline packages/ripple-binary-codec/src/enums/definitions.json:2788]: This changes all the numbers, which isn't backwards-compatible.

Comment (mvadari) [reply packages/ripple-binary-codec/src/enums/definitions.json:2788]: Auto-generating _is_ sufficient, I'm saying this is a rippled problem.

Comment (mvadari) [reply packages/ripple-binary-codec/src/enums/definitions.json:2788]: This is only a problem if you plan to merge this PR in as-is - if it's just for a beta version or something, it doesn't really matter.

File: packages/xrpl/test/models/payment.test.ts
Comment (khancode) [inline packages/xrpl/test/models/payment.test.ts:274]: Can you add an equivalent integration test in `test/integration/transactions/payment.test.ts`?

File: packages/xrpl/test/integration/transactions/mptokenAuthorize.test.ts
Comment (khancode) [inline packages/xrpl/test/integration/transactions/mptokenAuthorize.test.ts:79]:
```suggestion
      accountObjectsResponse = await testContext.client.request({
        command: 'account_objects',
        account: wallet2.classicAddress,
        type: 'mptoken',
      })
```
To ensure it's an `MPToken` ledger object.

File: packages/xrpl/test/integration/transactions/mptokenAuthorize.test.ts
Comment (khancode) [inline packages/xrpl/test/integration/transactions/mptokenAuthorize.test.ts:107]:
```suggestion
      accountObjectsResponse = await testContext.client.request({
        command: 'account_objects',
        account: wallet2.classicAddress,
        type: 'mptoken',
      })
```
ditto

File: packages/xrpl/src/models/transactions/MPTokenIssuanceSet.ts
Comment (khancode) [inline packages/xrpl/src/models/transactions/MPTokenIssuanceSet.ts:27]: This is duplicated

**Comment by justinr1234** [comment]
@coderabbit full review

**Comment by justinr1234** [comment]
@coderabbitai full review

**Review by khancode — APPROVED** [review]
LGTM

**Review by khancode — CHANGES_REQUESTED** [review]
The integration tests fail because it's not using a rippled docker image version with MPT enabled. Can you add it to resolve the tests or just comment out the tests for now until it's available?

File: packages/xrpl/src/models/transactions/MPTokenIssuanceSet.ts
Comment (achowdhry-ripple) [inline packages/xrpl/src/models/transactions/MPTokenIssuanceSet.ts:75]: Can use the isFlagEnabled util function here

File: packages/xrpl/src/models/transactions/MPTokenIssuanceSet.ts
Comment (achowdhry-ripple) [inline packages/xrpl/src/models/transactions/MPTokenIssuanceSet.ts:70]: eslint disable can be removed when isFlagEnabled is used

File: packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts
Comment (achowdhry-ripple) [inline packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts:135]: nit: we should probably make this number a constant for clarity

File: packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts
Comment (achowdhry-ripple) [inline packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts:7]: nit: can we rename this to indicate that its checking for a number format, and move it to the utils?

File: packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts
Comment (achowdhry-ripple) [inline packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts:114]: should validate the optional fields using the util helper function

File: packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts
Comment (achowdhry-ripple) [reply packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts:102]: I agree with this, but can use the helper functions for a bunch of the logic here. Some will need to be custom still (non negative asset scale, etc)

File: packages/xrpl/src/models/common/index.ts
Comment (achowdhry-ripple) [inline packages/xrpl/src/models/common/index.ts:28]: Could we Add MPTAmount here, since we group MPT and IOU typings in the isAmount check later on anyways? This way you would avoid the Amount | MPTAmount changes in a lot of the transactions

**Comment by achowdhry-ripple** [comment]
Quick note: you can refer to the Verified Credentials PR to see the CI docker changes needed to get the integration tests to run. Once you switch to 2.3.0+, there is an entrypoint issue that requires a manual override in the script, so you will have to make the changes here too

File: packages/xrpl/src/client/partialPayment.ts
Comment (achowdhry-ripple) [inline packages/xrpl/src/client/partialPayment.ts:36]: can also be simplified if MPTAmount is a part of Amount

File: packages/xrpl/HISTORY.md
Comment (achowdhry-ripple) [inline packages/xrpl/HISTORY.md:47]: This should probably still follow the exisiting versioning, so 3.0.0

**Review by achowdhry-ripple — CHANGES_REQUESTED** [review]
LGTM functionally, just a few code cleanups

File: packages/xrpl/src/models/common/index.ts
Comment (shawnxie999) [reply packages/xrpl/src/models/common/index.ts:28]: Thanks for noting this! This choice was chosen intentionally (although I agree it's not the most ideal). The reason why MPTAmount isn't included because currently only a few transactions support MPTAmount (Payment, Clawback, etc), while other transactions (like CheckCreate) doesn't. This is why `Amount | MPTAmount` is seen for certain fields.

Comment (shawnxie999) [reply packages/xrpl/src/models/common/index.ts:28]: We should probably create a ticket to remind to update it when MPTAmount gets incorporated into all the features

File: packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts
Comment (achowdhry-ripple) [reply packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts:114]: You could use isString instead (like oracleSet does for example)

` validateOptionalField(tx, 'maximumAmount', isString)`

File: packages/xrpl/src/models/common/index.ts
Comment (achowdhry-ripple) [reply packages/xrpl/src/models/common/index.ts:28]: Sure makes sense. Could you add an issue in the repo so we can track this for reference

**Comment by achowdhry-ripple** [comment]
Seeing the integration tests failing now, will need to be resolved. Looks like they are running now so the docker part is good

**Comment by achowdhry-ripple** [comment]
nit: I also agree with the typos raised by coderabbit, shoudl be quick fixes

File: packages/xrpl/test/models/MPTokenAuthorize.test.ts
Comment (shawnxie999) [reply packages/xrpl/test/models/MPTokenAuthorize.test.ts:18]: not needed

File: packages/xrpl/test/models/MPTokenAuthorize.test.ts
Comment (shawnxie999) [reply packages/xrpl/test/models/MPTokenAuthorize.test.ts:71]: we don't need to test these in the library

File: packages/xrpl/test/integration/transactions/mptokenAuthorize.test.ts
Comment (shawnxie999) [reply packages/xrpl/test/integration/transactions/mptokenAuthorize.test.ts:118]: we don't generally need to test rippled behaviors in client library, because it should be covered by integration test and unit test for rippled already

File: packages/xrpl/test/integration/transactions/clawback.test.ts
Comment (shawnxie999) [reply packages/xrpl/test/integration/transactions/clawback.test.ts:178]: not needed

**Comment by shawnxie999** [comment]
@achowdhry-ripple fixed comments from coderabbits that are relevant. previously the integration test failed with temDISABLED because the rippled cfg didn't include the MPT amendment.

File: .ci-config/rippled.cfg
Comment (khancode) [inline .ci-config/rippled.cfg:188]:
```suggestion
# 2.3.0-rc1 Amendments
fixAMMv1_1
Credentials
NFTokenMintOffer
MPTokensV1
fixNFTokenPageLinks
fixInnerObjTemplate2
fixEnforceNFTokenTrustline
fixReducedOffersV2
```

File: .github/workflows/nodejs.yml
Comment (khancode) [inline .github/workflows/nodejs.yml:111]: Why change this command when `RIPPLED_DOCKER_IMAGE` should ideally only be updated?

Comment (shawnxie999) [reply .github/workflows/nodejs.yml:111]: I copied what @achowdhry-ripple did in the credential PR.

File: packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts
Comment (khancode) [inline packages/xrpl/src/models/transactions/MPTokenIssuanceCreate.ts:88]:
```suggestion
   * It is a non-negative integer string that can store a range of up to 63 bits. If not set, the max
```

This will help avoid confusion of setting a raw integer type instead of an integer string.

**Review by khancode — APPROVED** [review]
LGTM

**Comment by tequdev** [comment]
MPToken is an asset with a balance, don't we need to change to `getBalanceChanges`(packages/xrpl/src/utils/getBalanceChanges.ts)?

**Comment by shawnxie999** [comment]
> MPToken is an asset with a balance, don't we need to change to `getBalanceChanges`(packages/xrpl/src/utils/getBalanceChanges.ts)?

Good catch! Yes, we can support this later separately.

**Comment by achowdhry-ripple** [comment]
> MPToken is an asset with a balance, don't we need to change to `getBalanceChanges`(packages/xrpl/src/utils/getBalanceChanges.ts)?

Also agree, this is a good catch. If its not too much overhead would say it should go out with this, otherwise we should update the JSDoc to say this does not include MPTs for now

**Comment by shawnxie999** [comment]
@achowdhry-ripple i would prefer doing it separately, may not be necessary to hold off merging the main change for a sub-component

**Comment by achowdhry-ripple** [comment]
> @achowdhry-ripple i would prefer doing it separately, may not be necessary to hold off merging the main change for a sub-component

In that case, could you raise a github issue here so we can track it?

**Pull request merged**
